### PR TITLE
Entity type identification

### PR DESCRIPTION
**Pull request by ahemphill**

### CODE REVIEW

File: tools/scoring/tests/fixtures.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+from score.parse_config import ParseConfig
+
+from validate import handler as validator
+from validate.generate_universe import BuildUniverse
+
+PROPOSED, SOLUTION = FileTypes
+SIMPLE, COMPLEX = DimensionCategories
+
+
+class TestFixtures:
```
Comment (charbull) [inline tools/scoring/tests/fixtures.py:26]: TestHelper?

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: This file is used as a test fixture; it is a valid DBO instance
+# which demonstrates the features of a reporting entity.
+
```
Comment (charbull) [inline tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:17]: why not reuse the files you defined in the previous PR ?
are they the same content?

File: tools/scoring/tests/dimensions/entity_type_identification_test.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        'tests/samples/proposed/entity_type_identification_reporting.yaml')
+    self.middling_score_argument = TestFixtures.prepare_dimension_argument(
+        entity_type=COMPLEX,
+        proposed_path=reporting_entity_file_path,
+        solution_path=featureful_file_path)
+
+  def testEvaluate_ScoreNone(self):
+    """When ceiling==0, the resulting score is None. The ceiling is 0
+    because the solution does not contain any entities."""
+    none_score_expected = EntityIdentification(
```
Comment (charbull) [inline tools/scoring/tests/dimensions/entity_type_identification_test.py:54]: shouldnt this be EntityType?

**Review by charbull — CHANGES_REQUESTED** [review]

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml (hunk 2 above)
Comment (ahemphill) [reply tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:17]: They are presently but I would prefer to leave them separate given that they are testing separate features; that way, there are no hidden interdependencies and modifications will not have unintended results.

File: tools/scoring/tests/dimensions/entity_type_identification_test.py (hunk 3 above)
Comment (ahemphill) [reply tools/scoring/tests/dimensions/entity_type_identification_test.py:54]: D'oh!

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml (hunk 2 above)
Comment (ahemphill) [reply tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:17]: Case in point: I have to change it to fix the test. :)

**Comment by ahemphill** [comment]
CI is now mysteriously failing on the instance validator; any ideas?

**Comment by charbull** [comment]
> CI is now mysteriously failing on the instance validator; any ideas?

it is a pylint issue, you can check the logs, it is an easy fix

**Comment by ahemphill** [comment]
I haven't touched the instance validator.

File: tools/scoring/score/dimensions/dimension.py [hunk 4]
```diff
@@ -337,14 +337,14 @@ def match_virtual_entities(
         none_subscore_reference = _VirtualEntityMatch(
             correct=0,
             correct_ceiling=len(solution_raw_field_names),
```
Comment (charbull) [inline tools/scoring/score/dimensions/dimension.py:339]: correct_ceiling and incorrect have the same values?

File: tools/scoring/score/dimensions/dimension.py [hunk 5]
```diff
@@ -337,14 +337,14 @@ def match_virtual_entities(
...
-            incorrect=0,
+            incorrect=len(solution_raw_field_names),
             proposed=set([]),
             solution=solution_parameters,
             types_correct=0,
             types_correct_ceiling=len(
```
Comment (charbull) [inline tools/scoring/score/dimensions/dimension.py:344]: same here. thats a bit odd to have two variables with the same values.

File: tools/scoring/score/dimensions/entity_type_identification.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    # if the field exists at the source
+    # add its raw field name to the set.
+    # Filter out sets which have no items
+    # and sort by number of fields represented in descending order.
+    # TODO: move the filtering/sorting functionality to add clarity.
+    return sorted(
+        list(
+            filter(
+                lambda tup: len(tup[0]) > 0,  # (raw field names, entity type)
+                [(set(file[link.source].translation[target_field].raw_field_name
```
Comment (charbull) [inline tools/scoring/score/dimensions/entity_type_identification.py:52]: can you please comment inline here, I dont follow what this is doing, it is hard to follow.

File: tools/scoring/score/dimensions/entity_type_identification.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        len(solution_types)
+        for proposed_types, solution_types in matches_reporting
+    ])
+    self.incorrect_reporting = sum([
+        len(solution_types.difference(proposed_types))
+        for proposed_types, solution_types in matches_reporting
+    ])
+
+  def evaluate(self):
+    """Calculates and assigns properties necessary
```
Comment (charbull) [inline tools/scoring/score/dimensions/entity_type_identification.py:158]: entity type identification?

File: tools/scoring/tests/dimensions/entity_type_identification_test.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+PROPOSED, SOLUTION = FileTypes
+SIMPLE, COMPLEX = DimensionCategories
+
+
+class EntityTypeIdentificationTest(absltest.TestCase):
+  def setUp(self):
+    super().setUp()
+    featureful_file_path = (
+        'tests/samples/proposed/entity_type_identification_virtual.yaml')
```
Comment (charbull) [inline tools/scoring/tests/dimensions/entity_type_identification_test.py:32]: if the same content please reuse the file.

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml (hunk 2 above)
Comment (charbull) [reply tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:17]: I propose if they are the same reuse the same, once you need to change the content then fork them.

**Review by charbull — CHANGES_REQUESTED** [review]

File: tools/scoring/score/dimensions/dimension.py (hunk 5 above)
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:344]: As above.

File: tools/scoring/score/dimensions/entity_type_identification.py (hunk 6 above)
Comment (ahemphill) [reply tools/scoring/score/dimensions/entity_type_identification.py:52]: It's commented above.

File: tools/scoring/tests/dimensions/entity_type_identification_test.py (hunk 8 above)
Comment (ahemphill) [reply tools/scoring/tests/dimensions/entity_type_identification_test.py:32]: It's no longer the same content.

File: tools/scoring/score/dimensions/dimension.py (hunk 4 above)
Comment (charbull) [reply tools/scoring/score/dimensions/dimension.py:339]: why not binding them together?
`incorrect = MAX_CORRECT - correct`?

File: tools/scoring/score/dimensions/entity_type_identification.py (hunk 6 above)
Comment (charbull) [reply tools/scoring/score/dimensions/entity_type_identification.py:52]: sure but I dont follow it maybe use a helper or apply the TODO?

File: tools/scoring/score/dimensions/entity_type_identification.py (hunk 7 above)
Comment (charbull) [reply tools/scoring/score/dimensions/entity_type_identification.py:158]:
```
calculates entity type identification scores for all devices.
```

File: tools/scoring/tests/dimensions/entity_identification_test.py [hunk 9]
```diff
@@ -18,73 +18,32 @@
 
 from score.dimensions.entity_identification import EntityIdentification
 from score.constants import FileTypes, DimensionCategories
-from score.parse_config import ParseConfig
 
-from validate import handler as validator
-from validate.generate_universe import BuildUniverse
+from ..helper import TestHelper
```
Comment (charbull) [inline tools/scoring/tests/dimensions/entity_identification_test.py:22]: please use the path as we transform the imports when we bring the code internally.

File: tools/scoring/tests/dimensions/entity_type_identification_test.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+# limitations under the License.
+"""Test for configuration file scoring tool
+"entity type identification" dimension (entity_type_identification.py)."""
+
+from absl.testing import absltest
+
+from score.dimensions.entity_type_identification import EntityTypeIdentification
+from score.constants import FileTypes, DimensionCategories
+
+from ..helper import TestHelper
```
Comment (charbull) [inline tools/scoring/tests/dimensions/entity_type_identification_test.py:22]: same here

**Review by charbull — CHANGES_REQUESTED** [review]

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+VLV-2:
+  cloud_device_id: 2599571827855502
+  guid: edfa7b3a-80da-4243-a95a-5ca24311ac45
+  id: CDM/196h$V6TP3C8vhrwG9nopt
```
Comment (charbull) [inline tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:21]: remove

File: tools/scoring/score/dimensions/dimension.py (hunk 4 above)
Comment (charbull) [reply tools/scoring/score/dimensions/dimension.py:339]: can you please add a todo to bind them together

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: This file is used as a test fixture; it is a valid DBO instance
+# which demonstrates the features of a reporting entity.
+
+VLV-2:
+  cloud_device_id: 2599571827855502
```
Comment (charbull) [inline tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:19]: non-blocking:
what happens when you provide the new format to the scorer?
```
edfa7b3a-80da-4243-a95a-5ca24311ac45:
  code : VLV-2
  translation:
    chilled_water_flowrate_sensor:
      present_value: points.chilled_water_flowrate_sensor.present_value
      units:
        key: pointset.points.chilled_water_flowrate_sensor.units
        values:
          liters_per_second: L/s
  type: HVAC/CHWS_WDT # Th
```

**Review by charbull — CHANGES_REQUESTED** [review]
just add the todos please

File: tools/scoring/score/dimensions/dimension.py (hunk 4 above)
Comment (ahemphill) [reply tools/scoring/score/dimensions/dimension.py:339]: I don't understand what you're asking: to reiterate our conversation on Friday, they are only the same value in this case. They are not inherently related. (See `:307` for another instance of `_VirtualEntityMatch` where the values differ.) Would this help?
```
correct_ceiling =len(solution_raw_field_names)
        none_subscore_reference = _VirtualEntityMatch(
            correct=0,
            correct_ceiling=correct_ceiling,
            incorrect=correct_ceiling,
```
Otherwise, a todo exists to pare down the attributes: https://trello.com/c/MBAeYiwI/31-pare-down-virtualentitymatch-attributes-following-implementation-of-entity-type-id

File: tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml (hunk 12 above)
Comment (charbull) [reply tools/scoring/tests/samples/proposed/entity_type_identification_reporting.yaml:19]:
```
edfa7b3a-80da-4243-a95a-5ca24311ac45:
  code : VLV-2
  cloud_device_id: 2599571827855502
  translation:
    chilled_water_flowrate_sensor:
      present_value: points.chilled_water_flowrate_sensor.present_value
      units:
        key: pointset.points.chilled_water_flowrate_sensor.units
        values:
          liters_per_second: L/s
  type: HVAC/CHWS_WDT # Th
  ```

**Review by charbull — APPROVED** [review]
